File: bridges/mcp-ts/src/discovery.ts
/**
 * ANIP service discovery and manifest fetching.
 */

export interface ANIPCapability {
  name: string;
  description: string;
  sideEffect: string;
  rollbackWindow: string | null;
  minimumScope: string[];
  financial: boolean;
  contractVersion: string;
  inputs: Array<{
    name: string;
    type: string;
    required?: boolean;
    default?: unknown;
    description?: string;
  }>;
  output: { type: string; fields: string[] };
  cost: Record<string, unknown> | null;
  requires: Array<{ capability: string; reason: string }>;
}

export interface ANIPService {
  baseUrl: string;
  protocol: string;
  compliance: string;
  endpoints: Record<string, string>;
  capabilities: Map<string, ANIPCapability>;
}

export async function discoverService(anipUrl: string): Promise<ANIPService> {
  // Step 1: Fetch discovery document
  const discoveryUrl = `${anipUrl.replace(/\/$/, "")}/.well-known/anip`;
  const discoveryResp = await fetch(discoveryUrl);
  if (!discoveryResp.ok) {
    throw new Error(
      `Discovery failed: ${discoveryResp.status} ${discoveryResp.statusText}`
    );
  }
  const { anip_discovery: discovery } = (await discoveryResp.json()) as {
    anip_discovery: Record<string, unknown>;
  };

  const baseUrl = discovery.base_url as string;
  const endpoints = discovery.endpoints as Record<string, string>;

  // Step 2: Fetch full manifest
  const manifestUrl = resolveUrl(baseUrl, endpoints.manifest);
  const manifestResp = await fetch(manifestUrl);
  if (!manifestResp.ok) {
    throw new Error(
      `Manifest fetch failed: ${manifestResp.status} ${manifestResp.statusText}`
    );
  }
  const manifest = (await manifestResp.json()) as {
    capabilities: Record<string, Record<string, unknown>>;
  };

  // Step 3: Build capability objects
  const capabilities = new Map<string, ANIPCapability>();
  for (const [name, cap] of Object.entries(manifest.capabilities)) {
    const sideEffect = cap.side_effect as { type: string; rollback_window?: string };
    const minimumScope = cap.minimum_scope as string[];
    capabilities.set(name, {
      name,
      description: cap.description as string,
      sideEffect: sideEffect.type,
      rollbackWindow: sideEffect.rollback_window ?? null,
      minimumScope: minimumScope,
      financial: (cap.cost as Record<string, unknown>)?.financial != null,
      contractVersion: (cap.contract_version as string) ?? "1.0",
      inputs: (cap.inputs as ANIPCapability["inputs"]) ?? [],
      output: (cap.output as ANIPCapability["output"]) ?? {
        type: "unknown",
        fields: [],
      },
      cost: (cap.cost as Record<string, unknown>) ?? null,
      requires:
        (cap.requires as Array<{ capability: string; reason: string }>) ?? [],
    });
  }

  // Resolve all endpoint URLs
  const resolvedEndpoints: Record<string, string> = {};
  for (const [k, v] of Object.entries(endpoints)) {
    resolvedEndpoints[k] = resolveUrl(baseUrl, v);
  }

  return {
    baseUrl,
    protocol: discovery.protocol as string,
    compliance: (discovery.compliance as string) ?? "anip-compliant",
    endpoints: resolvedEndpoints,
    capabilities,
  };
}

function resolveUrl(baseUrl: string, path: string): string {
  if (path.startsWith("http")) return path;
  return `${baseUrl.replace(/\/$/, "")}/${path.replace(/^\//, "")}`;
}

File: bridges/mcp/anip_mcp_bridge/discovery.py
"""ANIP service discovery and manifest fetching."""

from __future__ import annotations

from dataclasses import dataclass, field

import httpx


@dataclass
class ANIPCapability:
    """A discovered ANIP capability with full metadata."""

    name: str
    description: str
    side_effect: str
    rollback_window: str | None
    minimum_scope: list[str]
    financial: bool
    contract_version: str
    inputs: list[dict]
    output: dict
    cost: dict | None
    requires: list[dict]


@dataclass
class ANIPService:
    """A discovered ANIP service."""

    base_url: str
    protocol: str
    compliance: str
    endpoints: dict[str, str]
    capabilities: dict[str, ANIPCapability]
    profiles: dict[str, str] = field(default_factory=dict)


async def discover_service(anip_url: str) -> ANIPService:
    """Discover an ANIP service from its base URL.

    Follows the standard ANIP flow:
    1. Fetch /.well-known/anip
    2. Fetch full manifest from the declared endpoint
    3. Return structured service description
    """
    async with httpx.AsyncClient(timeout=30) as client:
        # Step 1: Discovery document
        discovery_url = f"{anip_url.rstrip('/')}/.well-known/anip"
        resp = await client.get(discovery_url)
        resp.raise_for_status()
        discovery = resp.json()["anip_discovery"]

        base_url = discovery["base_url"]
        endpoints = discovery["endpoints"]

        # Step 2: Full manifest
        manifest_url = _resolve_url(base_url, endpoints["manifest"])
        resp = await client.get(manifest_url)
        resp.raise_for_status()
        manifest = resp.json()

        # Step 3: Build capability objects from manifest
        capabilities = {}
        for name, cap in manifest["capabilities"].items():
            capabilities[name] = ANIPCapability(
                name=name,
                description=cap["description"],
                side_effect=cap["side_effect"]["type"],
                rollback_window=cap["side_effect"].get("rollback_window"),
                minimum_scope=cap.get("minimum_scope", cap.get("required_scope", [])),
                financial=cap.get("cost", {}).get("financial") is not None,
                contract_version=cap.get("contract_version", "1.0"),
                inputs=cap.get("inputs", []),
                output=cap.get("output", {}),
                cost=cap.get("cost"),
                requires=cap.get("requires", []),
            )

        # Resolve all endpoint URLs
        resolved_endpoints = {
            k: _resolve_url(base_url, v) for k, v in endpoints.items()
        }

        return ANIPService(
            base_url=base_url,
            protocol=discovery["protocol"],
            compliance=discovery.get("compliance", "anip-compliant"),
            endpoints=resolved_endpoints,
            capabilities=capabilities,
            profiles=discovery.get("profile", {}),
        )


def _resolve_url(base_url: str, path: str) -> str:
    """Resolve a relative endpoint path against the base URL."""
    if path.startswith("http"):
        return path
    return f"{base_url.rstrip('/')}/{path.lstrip('/')}"
